src/main/java/com/fasterxml/jackson/databind/node/ObjectNode.java

```diff
@@ -342,44 +344,50 @@ public JsonNode findValue(String propertyName)
     @Override
     public List<JsonNode> findValues(String propertyName, List<JsonNode> foundSoFar)
     {
-        for (Map.Entry<String, JsonNode> entry : _children.entrySet()) {
-            if (propertyName.equals(entry.getKey())) {
-                if (foundSoFar == null) {
-                    foundSoFar = new ArrayList<JsonNode>();
-                }
-                foundSoFar.add(entry.getValue());
-            } else { // only add children if parent not added
-                foundSoFar = entry.getValue().findValues(propertyName, foundSoFar);
+        JsonNode jsonNode = _children.get(propertyName);
+        if (jsonNode != null) {
+            if (foundSoFar == null) {
+                foundSoFar = new ArrayList<>();
             }
+            foundSoFar.add(jsonNode);
+            return foundSoFar;
```

Comment:
And here's the problem that was reported as #4229: we are to collect ALL matches, not just main-level one